Optimizations + target DNS injection

deputinizer opened the pull request:
Currently looking at pprof, more changes coming.

deputinizer committed:
fix: make "too fast TCP SYN (dial spam)" less noticeable

deputinizer committed:
perf: optimize mallocgc in customresolver.GetIPs

deputinizer committed:
Merge remote-tracking branch 'upstream/main' into fix-optimize

deputinizer committed:
fix: merge

baylanger commented on internal/stoppropaganda/customresolver/customgetip.go:
```diff
@@ -0,0 +1,47 @@
+		for i := 0; i < len(ips); i++ {
+			ip := ips[i]
+			if ipv4 := ip.To4(); ipv4 == nil {
+				// Remove inplace trick
```
Thank you for putting some comments, trying to learn go ;)

deputinizer replied:
Yeah, that's a neat trick

deputinizer commented:
<details>
  <summary>Still, dial is the main CPU eater</summary>

![image](https://user-images.githubusercontent.com/100740281/156802575-6ac6e8a6-2f47-43fe-804a-f2c62090e6a0.png)

</details>

deputinizer committed:
fix: first check concurrency, then limit TCP SYN
Works a lot better

deputinizer committed:
fix: free memory of IP []byte slice

deputinizer commented:
<details>
  <summary>[Spoiler] Memory usage from pprof</summary>
  
![image](https://user-images.githubusercontent.com/100740281/156803854-2a6ad720-cc08-4283-98e6-da9908e6530e.png)

</details>

deputinizer committed:
feat: targets in one package

deputinizer committed:
feat: inject DNS servers from dns.go

deputinizer committed:
feat: use injection DNS resolver

deputinizer committed:
feat: use injection DNS resolver

deputinizer committed:
feat: first dial Yandex/reliable DNS, then localhost resolver

deputinizer committed:
refactor: raname DNSServer -> DNSTarget to avoid confusion

deputinizer committed:
feat: add github.com/valyala/fastrand

deputinizer committed:
feat: add github.com/valyala/fastrand vendor

deputinizer committed:
perf: faster dns randomization

deputinizer committed:
fix: randomize in every DNS request

deputinizer commented:
Why there was no `getRandomDomain()` in every DNS request?

deputinizer committed:
refactor: http.go renamed to apihttp.go

deputinizer committed:
feat: add /dnscache debug in apihttp.go

deputinizer committed:
fix: revert fasthttp.ReleaseResponse
Program uses a lot less memory after few minutes.
Keeping response object and holding body buffer was bad idea i suppose.

deputinizer commented:
Yeah, now it runs constant 100 Mbit/s :100:

erkexzcx commented:
Good to be merged?

erkexzcx commented:
> Why there was no `getRandomDomain()` in every DNS request?

Because it was CPU waste which does not bring any benefit. I did not want to use `russiagofuckyourself.ru` or something like that, since it would make very easy to filter out such requests if they want to check their logs. Creating a random domain once during startup would be sufficient IMO.

deputinizer commented:
Works, so far it didn't panic :+1:

deputinizer commented:
> Creating a random domain once during startup would be sufficient IMO.

But they can cache the 404 response and return instantly.

deputinizer commented:
> Good to be merged?

Oh, let's also reuse the `fastrand.RNG`. Wait a minute

deputinizer committed:
perf: reuse fastrand.RNG

erkexzcx commented:
> But they can cache the 404 response and return instantly.

ohhh that makes sense. Randomizing domains actually helps then.

deputinizer commented:
All done.

erkexzcx commented:
Let me do some tests and merge.

erkexzcx commented:
P.S. Can you try to fix this? https://github.com/erkexzcx/stoppropaganda/issues/100

EDIT: Doesn't seem to appear with your PR code.

erkexzcx commented:
Good job

erkexzcx commented:
With this PR it even started to work on Android. Perfection...
